### ISSUE TITLE
Bug: floating point rounding discrepancies on some architectures

### ISSUE THREAD

**Issue by tillea** [taoliu/MACS#359]
**Describe the bug**
The Debian packaged version of MACS has shown floating point rounding issues when building on different hardware architectures.  You can find the affected architectures on the [autobuilder page](https://buildd.debian.org/status/package.php?p=macs).

There is a [bug report in the Debian bug tracking system](https://bugs.debian.org/949204) explaining the issue.  The bug report extracts the main information from these logs:

    > 14.9.6 checking callpeak_pe_broad run_callpeak_bedpe_broad_peaks.broadPeak ...
    >   ... failed! Difference:
    > 4c4
    > < chr22 17398268 17398740 run_callpeak_bedpe_broad_peak_4 279 . 13.95538 30.73891 27.95395
    > ---
    > > chr22 17398268 17398740 run_callpeak_bedpe_broad_peak_4 279 . 13.95538 30.73891 27.95394

    i386, mips64el and mipsel, failed with a NaN verses 0 discrepancy

    > E       0.0
    > E
    > E       - (-48.6784, nan)
    > E       ?            ^^^
    > E
    > E       + (-48.6784, 0.0)
    > E       ?            ^^^
    and s390x failed with a zerodivisionerror

    >    File "MACS2/IO/Parser.pyx", line 1153, in MACS2.IO.Parser.BAMPEParser.build_petrack
    > ZeroDivisionError: float division
    >   ... error/warning found!

**To Reproduce**
I tried to reproduce the issue in an i386 chroot - but here the test suite worked (may be due to the fact that amd64 floating point is used in this case.  So testing on a real hardware emulation or real hardware is necessary.

**Expected behavior**
The test suite should run on all supported hardware platforms.

**System (please complete the following information):**
 - OS: Debian GNU Linux - different hardware
 - Python version: 3.7.5
 - Numpy version: 1.17.4
 - MACS Version: 2.2.6

Kind regards, Andreas.

**Comment by mr-c** [comment]
see also #351 #349 and maybe #340

**Comment by taoliu** [comment]
Thanks for the information and let's keep track of this issue! I think I have to find some other way to test these. Currently, the only tool I have is TravisCI which has limited multi-architecture supports/hardware emulation. I think it only supports amd64 and arm64?

**Comment by mr-c** [comment]
@taoliu Travis supports 4 architectures (without emulation): here's a PR to add them all https://github.com/taoliu/MACS/pull/360

https://docs.travis-ci.com/user/multi-cpu-architectures/

**Comment by taoliu** [comment]
@mr-c Thanks! But Travis doesn't support actual i386. How does Debian build system support so many architectures? Is there real hardware behind?

**Comment by mr-c** [comment]
@taoliu Yes, Debian builds with real hardware. They have over 40 build servers and numerous porterboxes: https://db.debian.org/machines.cgi

I can also add a hack to force 32bit compilation for Travis

**Comment by mr-c** [comment]
Maybe @junaruga has a Travis 32bit Python recipe handy?

**Comment by junaruga** [comment]
Yes, it's possible. Intel 32-bit? Or ARM 32-bit?

**Comment by mr-c** [comment]
@junaruga 32bit x86, please and thank you!

**Comment by junaruga** [comment]
@mr-c I sent PR to show the example to add the x86 32-bit to Travis. https://github.com/taoliu/MACS/pull/361

In this case, we can use i386 binary container + QEMU rather than running `setarch` and `foo:i386` libraries (= [example](https://github.com/junaruga/ci-multi-arch-native-test/blob/master/.travis.yml#L19-L31)).

**Comment by taoliu** [comment]
@junaruga Thank you!

**Comment by tillea** [comment]
Any news?

**Comment by taoliu** [comment]
@tillea I am still working on it.

**Comment by taoliu** [comment]
Solved the problems in most of the architectures supported by TravisCI by replacing Python int/float types in pyx files with more specific NumPy intXX_t/floatXX_t versions. However, my current implementation to read the BAM file can't work on s390x. It seems that MACS2 can't correctly read the bwflag (bit-wise flag) in BAMPE file, for example, in the following code, `bwflag` can be `163` in other arches, but be `41728` in s390x. 

https://github.com/taoliu/MACS/blob/36492d89877f647375799d93a588cb1a0b6fbab0/MACS2/IO/Parser.pyx#L1230-L1240

I didn't see such issue in BAMParser which read the BAM file in a similar way. And BAM should be compressed in little-endian format regardless. I will explore the solution when I get some time. 

See the pull request: PR #367

**Comment by junaruga** [comment]
> for example, in the following code, bwflag can be 163 in other arches, but be 41728 in s390x.

s390x is big-endian. other architectures amd64 (x86_64), i386, arm64, ppc64le are big-endian.
It might be big-endian specific issue.

**Comment by taoliu** [comment]
@junaruga I guess so, however, the BAMParser uses the same code and I don't see any error message.  Quite strange. I may need to do more tests.

https://github.com/taoliu/MACS/blob/36492d89877f647375799d93a588cb1a0b6fbab0/MACS2/IO/Parser.pyx#L1048-L1058

**Comment by junaruga** [comment]
The cast from data (`unsigned char *`) to ui16 (`uint16_t *`) might be failed in big-endian?
What happens when using `ui8` (`uint8_t`) instead of ui16 (`uint16_t *`)?

```
ui16 = <uint16_t *>data 
```

https://www.digital-detective.net/understanding-big-and-little-endian-byte-order/

**Comment by taoliu** [comment]
@junaruga Yes. Hopefully, it's fixed in the PR #370 :)

**Comment by junaruga** [comment]
@taoliu Great!
